astropy/coordinates/stokes_coord.py

```diff
@@ -0,0 +1,160 @@
+
+    original_mapping = copy(STOKES_VALUE_SYMBOL_MAP)
+    if not replace:
+        STOKES_VALUE_SYMBOL_MAP = {**original_mapping, **mapping}  # noqa: PLW0603
+    else:
+        STOKES_VALUE_SYMBOL_MAP = mapping  # noqa: PLW0603
+
+    yield
+
+    STOKES_VALUE_SYMBOL_MAP = original_mapping
+
+
+class StokesCoord(ShapedLikeNDArray):
+    """
+    A representation of stokes coordinates with helpers for converting to profile names.
+
+    Parameters
+    ----------
+    stokes_values : array-like
+        The numeric values representing stokes coordinates.
+    """
+
+    def __init__(self, stokes, copy=False):
+        stokes = np.asanyarray(stokes)
+        if stokes.dtype.kind == "U":
+            self._data = self._from_symbols(stokes)
+        else:
+            self._data = stokes
+
+    @property
+    def shape(self):
+        return self._data.shape
+
+    @property
+    def value(self):
+        return self._data
+
+    def __array__(self):
+        return self._data
+
+    def _apply(self, method, *args, **kwargs):
+        cls = type(self)
+
+        if callable(method):
+            return cls(method(self._data, *args, **kwargs))
+        elif method == "__getitem__":
+            return cls(self._data.__getitem__(*args))
+
+    @staticmethod
+    def _from_symbols(symbols):
+        """
+        Construct a StokesCoord from strings representing the stokes symbols.
+        """
+        values_array = np.full_like(symbols, np.nan, dtype=float, subok=False)
+        for stokes_value, symbol in STOKES_VALUE_SYMBOL_MAP.items():
+            values_array[symbols == symbol.symbol] = stokes_value
+
+        if (nan_values := np.isnan(values_array)).any():
+            raise ValueError(
+                f"Unknown stokes symbols present in the input array: {np.unique(symbols[nan_values])}"
+            )
+
+        return values_array
+
+    @property
+    def _stokes_values(self):
+        """
+        A representation of the coordinate as integers.
+        """
+        # Note we unbroadcast and re-broadcast here to prevent the new array
+        # using more memory than the old one.
+        return np.broadcast_to(np.round(unbroadcast(self._data)), self.shape)
+
+    @property
+    def symbol(self):
+        """
+        The coordinate represented as strings
+        """
+        known_symbols = tuple(
+            ["?"] + [s.symbol for s in STOKES_VALUE_SYMBOL_MAP.values()]
+        )
+        max_len = np.max([len(s) for s in known_symbols])
+
+        # Note we unbroadcast and re-broadcast here to prevent the new array
+        # using more memory than the old one.
+        symbolarr = np.full(unbroadcast(self._data).shape, "?", dtype=f"<U{max_len}")
+
+        for value, symbol in STOKES_VALUE_SYMBOL_MAP.items():
+            symbolarr[unbroadcast(self._stokes_values) == value] = symbol.symbol
+
+        return np.broadcast_to(symbolarr, self.shape)
+
+    def __eq__(self, other):
+        other = np.asanyarray(other)
```

Comment:
Would it make sense to do something like
```
try:
    other = self.__class__(other)
except Exception:
    return NotImplemented

return self._data == other._data
```

Reply:
I want to allow people to compare to the symbol strings, not just the numeric values. So I would like all of:

```python
>>> StokesCoord(["I", "Q"]) == "I"
>>> StokesCoord(["I", "Q"]) == ["Q", "V"]
>>> StokesCoord(["I", "Q"]) == [0, 1]
```

to work.

Reply:
As long asl those are valid input for `StokesCoord`, those should all be fine. It doesn't seem there should be a need to duplicate the logic of deciding between strings and integer.